### PR TITLE
Convert chart-controller to TypeScript

### PR DESCRIPTION
**Pull request by DanGonite57**
### Description

Convert chart-controller to TypeScript.

Continues from #2735 

Part of #2431

### CODE REVIEW

**Comment by Ferotiq** [comment]
https://www.npmjs.com/package/@types/chartjs-plugin-trendline

File: frontend/package-lock.json [hunk 1]
```diff
@@ -3646,16 +3665,6 @@
         "node": ">=8"
       }
     },
-    "node_modules/bindings": {
```
Comment (Bruception) [inline frontend/package-lock.json:3649]: This is no good.

**Review by Bruception — CHANGES_REQUESTED** [review]
Merge was incorrect. Please fix

**Comment by DanGonite57** [comment]
Pending merge of https://github.com/DefinitelyTyped/DefinitelyTyped/pull/59391

File: frontend/src/scripts/test/result.ts [hunk 2]
```diff
@@ -177,12 +176,9 @@ export async function updateGraphPBLine(): Promise<void> {
...
   }
-  ChartController.result.options.scales!["wpm"]!.max = Math.round(
-    maxChartVal + 5
-  );
-  ChartController.result.options.scales!["raw"]!.max = Math.round(
-    maxChartVal + 5
-  );
+  resultScaleOptions["wpm"].max = Math.round(maxChartVal);
+  resultScaleOptions["raw"].max = Math.round(maxChartVal);
+  ChartController.result.updateColors();
```
Comment (Miodec) [inline frontend/src/scripts/test/result.ts:181]: Please bring back the +5

File: frontend/src/scripts/test/result.ts [hunk 3]
```diff
@@ -177,12 +176,9 @@ export async function updateGraphPBLine(): Promise<void> {
...
   }
-  ChartController.result.options.scales!["wpm"]!.max = Math.round(
-    maxChartVal + 5
-  );
-  ChartController.result.options.scales!["raw"]!.max = Math.round(
-    maxChartVal + 5
-  );
+  resultScaleOptions["wpm"].max = Math.round(maxChartVal + 5);
+  resultScaleOptions["raw"].max = Math.round(maxChartVal + 5);
+  ChartController.result.updateColors();
```
Comment (Miodec) [inline frontend/src/scripts/test/result.ts:181]: Please remove

File: frontend/src/scripts/test/result.ts [hunk 4]
```diff
@@ -36,10 +45,9 @@ async function updateGraph(): Promise<void> {
       labels.push(i.toString());
     }
   }
-  (ChartController.result.data.labels as string[]) = labels;
-  (ChartController.result as Chart<"line" | "scatter">).options.scales![
-    "wpm"
-  ]!.title!.text = Config.alwaysShowCPM
+  ChartController.result.updateColors();
```
Comment (Miodec) [inline frontend/src/scripts/test/result.ts:48]: Another merge error

**Review by Miodec — CHANGES_REQUESTED** [review]
Those updates were removed because there were being called all at once for no reason - they were moved to `result.ts:631`

Other than that looks ok to me, but ill wait for another review from the boys that know more about typescript

File: frontend/src/scripts/test/result.ts (hunk 3 above)
Comment (DanGonite57) [reply frontend/src/scripts/test/result.ts]: Fixed

File: frontend/src/scripts/test/result.ts (hunk 4 above)
Comment (DanGonite57) [reply frontend/src/scripts/test/result.ts]: Fixed

File: frontend/src/scripts/types/types.d.ts [hunk 5]
```diff
@@ -128,6 +128,32 @@ declare namespace MonkeyTypes {
     | CustomLayoutFluid
     | `${string} ${string} ${string}`;
 
+  type HistoryChartData = {
+    x: number;
```
Comment (Ferotiq) [inline frontend/src/scripts/types/types.d.ts:132]: use interfaces

Comment (DanGonite57) [reply frontend/src/scripts/types/types.d.ts:132]: Fixed

**Review by Miodec — APPROVED** [review]

**Pull request merged**